projects/ti_to_ti/pre_tokenize_parallel_example.py

```diff
@@ -0,0 +1,319 @@
+    tokenizer = AutoTokenizer.from_pretrained(model_path)
+    local_output_paths = []
+    max_length = 0  # To track the maximum length
+    num_img = 0
+    num_text = 0
+    for piece in tqdm(input_data, desc=f"Processing on GPU {gpu}"):
+        formatted_sample = format_sample(piece)
+        preprocessed_sample, label_len, flag = preprocess(tokenizer, processor, formatted_sample)
+        if flag:
+            num_img += 1
+        else:
+            num_text += 1
+        # torch.set_printoptions(threshold=torch.inf)
+        
+        if preprocessed_sample['pixel_values'] is not None:
+            updated_piece = model.pre_tokenization(
+                input_ids=preprocessed_sample['input_ids'], 
+                pixel_values=preprocessed_sample['pixel_values'].to(device = model.device, dtype = torch.bfloat16),
+            )
+        else:
+            updated_piece = model.pre_tokenization(
+                input_ids=preprocessed_sample['input_ids']
+            )
+        updated_piece['labels'] = updated_piece['input_ids'].clone()
+        updated_piece['labels'][:label_len] = IGNORE_INDEX
+        if updated_piece['input_ids'].shape[0] <= 4096:
+            for key, value in updated_piece.items():
+                if torch.is_tensor(value):
+                    updated_piece[key] = value.cpu()
+            file_name = str(uuid.uuid4()) + '.pt'
+            file_path = os.path.join(cache_dir, file_name)
+            torch.save(updated_piece, file_path)  # Save the tensor data to file in CPU format
+            local_output_paths.append(file_path)  # Store the path for future reference
+            
+        # Clean up memory
+        del updated_piece
+        torch.cuda.empty_cache()
+
+    output_path.extend(local_output_paths)
+    print(f"GPU {gpu} processed {len(local_output_paths)} messages")
+
+def main():
+    input_path = "input.json"
+    output_path = "output.pt"
+    
+    model_path = "path_to_model"
+    cache_path = ".cache"
+    
+    # if cache dir does not exist, make one
+    if not os.path.exists(cache_path):
+        os.makedirs(cache_path)
+    
+    num_processes = 16
+    num_gpus = 8
+    output_paths = mp.Manager().list()  # For collecting results from multiple processes
+    
+    with open(input_path, 'r') as f:
+        input_data = json.load(f)
+    
+    target = input_data
+    chunks = [target[i::num_processes] for i in range(num_processes)]
+        
+    processes = []
+    for id in range(num_processes):
+        gpu = id % num_gpus  # This maps process to GPU cyclically
+        p = mp.Process(target=process_data, args=(gpu, chunks[id], model_path, output_paths, ".cache"))
+        p.start()
+        processes.append(p)
+
+    for p in processes:
+        p.join()
+    
+    output_paths = list(output_paths)
+    
+    all_data = []
+    for path in output_paths:
+        data = torch.load(path)
+        all_data.append(data)  # 假设每个文件中的数据本身就是一个列表
```

Comment:
remove the note

Reply:
fixed in the https://github.com/PKU-Alignment/align-anything/pull/41/commits/63f0f76733e3ca55563bffdf1fcb0512364a70ff commit